### ISSUE TITLE
Expensify.cash - Implement Higher-Order Component that re-renders when the locale changes

### ISSUE THREAD

**Issue by MitchExpensify** [Expensify/Expensify.cash#2176]
[Upwork job posting](https://www.upwork.com/jobs/~01211766e30f619205) 

We are implementing the structure for internationalisation and localisation in Expensify.Cash. As part of that, we need to implement a Higher-Order Component that re-renders when the locale changes

This is similar to our withOnyx HOC. It’s the only place that needs to care when the preferredLocale NVP changes and when it does, it re-triggers rendering of any component using it. It will:

1. Have one prop preferredLocale and its default will be `en`
2. Call withOnyx passing `{locale: {key: ONYXKEYS.PREFERRED_LOCALE}}`. ONYXKEYS.PREFERRED_LOCALE will have the value `preferredLocale`
3. Do I need to handle anything in componentDidUpdate? I think not?
4. It will pass as props to the WrappedComponent (the component that’s using it) each of the methods of the libraries defined above, but with the locale already passed (we can also make this just one property with all these methods in it. Something like `translations.methodName()`):

- translate(phrase, variables) which will use [this](https://github.com/Expensify/Expensify.cash/blob/b3172945b0ff4d9a0a223ec574706d1ed9601f40/src/libs/translate.js)
- numberFormat(options). It will return a new Intl.NumberFormat object for the current locale and using the options passed.
- timestampToRelative(timestamp) which will use [this](https://github.com/Expensify/Expensify.cash/blob/f579946fbfbdc62acc5bd281dc75cabb803d9af0/src/libs/DateUtils.js#L95) 
- timestampToDateTime(timestamp, includeTimezone) which will use [this](https://github.com/Expensify/Expensify.cash/blob/f579946fbfbdc62acc5bd281dc75cabb803d9af0/src/libs/DateUtils.js#L96)
- toLocalPhone(number) which will use [this](https://github.com/Expensify/Expensify.cash/blob/bdfbafe18ee2d60f766c697744f23fad64b62cad/src/libs/LocalePhoneNumber.js#L51)
- fromLocalPhone(number) which will use [this](https://github.com/Expensify/Expensify.cash/blob/bdfbafe18ee2d60f766c697744f23fad64b62cad/src/libs/LocalePhoneNumber.js#L52)

Expensify/Expensify repo [issue](https://github.com/Expensify/Expensify/issues/152152).

**Comment by parasharrajat** [comment]
I would like to work on this issue.
## Proposal
1. I will create A simple HOC under `components/withLocalize.js` (name could change to reflect the meaning more accurately) with one prop `preferredLocale` (defaults to `en`).
2. This HOC will listen for ONYX key `ONYXKEYS.PREFERRED_LOCALE`. Like this
```javascript
export default withOnyx({
    preferredLocale: {
        key: ONYXKEYS.PREFERRED_LOCALE
    },
})(withLocalize);
```
3. It would be simple Functional HOC(No need of class comoponent as we are not managing any state) that will pass down all the curried Localized functions mentioned in the Issue description to the `wrappedComponent`.   
``` javascript
const translations = {
                translate: passLocale(props.preferredLocale, translate),
                numberFormat : passLocale(props.preferredLocale, numberFormat),
                timestampToRelative: ...
                timestampToDateTime: ...
                toLocalPhone: ...
                fromLocalPhone: ...
            }
            return (
                <WrappedComponent
                    // eslint-disable-next-line react/jsx-props-no-spreading
                    {...this.props}
                    translations={translations}
                />
            );
            ................... // function which return a curried function passing locale as first argument.
            // It would be placed outside the HOC component.
            const passLocale = (locale, func) => {
               return (...props) => func(locale, ...props);
             };
 ```  
 4. We can either use a function like passLocale to get the curried function or just create a callback on the spot in the translations object. like
 ```javascript
 numberFormat : (options) => numberFormat(this.props.locale, options),
 ```

**Comment by iwiznia** [comment]
Proposal sounds good, go ahead and start working on the PR @parasharrajat 

> We can either use a function like passLocale to get the curried function or just create a callback on the spot in the translations object. like

I think doing it on the spot is better, since it's simpler to read/understand

**Comment by parasharrajat** [comment]
Quick Question: 

for `numbeFormat` you asked you use `Intl.numberformat`. but `Intl` will not be available on native devices. Are there any plans to support that on native platforms? 

I suggest using the polyfill https://formatjs.io/docs/polyfills/intl-numberformat/. We should create separate implementations of `Intl.numberformat` for web(available already) and native platforms.

**Comment by iwiznia** [comment]
Great question. I learned that [today](https://expensify.slack.com/archives/C01GTK53T8Q/p1617296293298100?thread_ts=1617294683.292800&cid=C01GTK53T8Q). I am not sure what is the best solution, kidroca there says there's a gradle config to enable Intl support. Is the polyfil better? Or can you summarize the pros and cons of each and which one you think is best?

**Comment by parasharrajat** [comment]
I am not an expert on the matter but here are few thoughts.

#### Why this issue on Android?

In order to build an app, React Native uses a virtual machine JavaScriptCore.

  - On iOS, it uses the JavaScriptCore provided by the iOS platform that already includes the toLocaleString() method.
  - On Android, it bundles the JavaScriptCore along with the application. In order to limit the app size, React Native uses a flavor that doesn’t include all JavaScript methods.

Now we have two options to format Numbers.
1. Use Intl Polyfill.
2. Change the JSC core which has Intl support. (this is where kidroca was saying the dramatic increase in App size).
 https://github.com/Expensify/Expensify.cash/blob/579459280a69dab87677925e603dd2967305ccd1/android/app/build.gradle#L108-L118
 
I think Polyfill will be very less in size. as opposed to JSC. I would request you to ask the same question with someone more experienced on the team as it's about a decision to increase the app size eventually.

**Comment by parasharrajat** [comment]
@iwiznia There are a couple of libs we can use for `numberFormat`. But I am not aware of the scope for the `numberFormat `function in the project.

http://numeraljs.com/
https://numbrojs.com/index.html
The Polyfill https://formatjs.io/docs/polyfills/intl-numberformat/.

**Comment by parasharrajat** [comment]
@roryabraham any thoughts here.

**Comment by iwiznia** [comment]
I asked around and people seem to prefer the polyfill too, to make a comparison, could you get the total size of both solutions please?

**Comment by AndrewGable** [comment]
BTW we use Hermes on Android: https://github.com/Expensify/Expensify.cash/pull/593 I think a polyfill or library will still be required, but just wanted to let you know.

**Comment by iwiznia** [comment]
Seems Hermes [does not support Intl](https://github.com/facebook/hermes/issues/23). So I guess we should go with the polyfills as that's our only option?
I am unsure which one though... seems that some people (1.5 years ago though) [think they are not really complete](https://github.com/facebook/hermes/issues/23#issuecomment-511019172)?

**Comment by parasharrajat** [comment]
@iwiznia. we have new polyfills. https://github.com/facebook/hermes/issues/23#issuecomment-605333778. They seem to support this. Anyways, we just need to polyfill. Intl.numberfomat for now. https://formatjs.io/docs/polyfills/intl-numberformat

**Comment by iwiznia** [comment]
Oh nice! ok, I think we should go with that option then. @AndrewGable, @Julesssss do you agree?

**Comment by AndrewGable** [comment]
Looks great 👍

**Comment by iwiznia** [comment]
@parasharrajat what's the state of this?

**Comment by parasharrajat** [comment]
@iwiznia I pushed a PR 5 days back. Waiting for review. #2208

**Comment by iwiznia** [comment]
Damn, seems it fell through the cracks... I will review it now.
